<a href="https://colab.research.google.com/github/RedZeptic/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/Thomas_Mineo_LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

- also try and practice some 1-sample t-tests

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)
3. You try and create some kind of a visualization that communicates the results of your hypothesis tests. This can be as simple as a histogram of the p-values or the t-statistics.

# After the fact...

#### please read.
not sure, I don't think null hypothisies belong with two sample t-tests correct? if so, I did this wrong.

In [0]:
# imports
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel, ttest_1samp
import numpy as np
import seaborn as sns
import pandas as pd

In [2]:
# getting the dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-16 20:28:23--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.04s   

2019-09-16 20:28:23 (461 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [2]:
# gotta load that data too
df = pd.read_csv('house-votes-84.data',
                 header=None,
                 names=['party', 'handicapped-infants', 'water-project',
                       'budget', 'physician-fee-freeze', 'el-salvador-aid',
                       'religious-groups', 'anti-satellite-ban',
                       'aid-to-contras', 'mx-missile', 'immigration',
                       'synfuels', 'education', 'right-to-sue', 'crime',
                       'duty-free', 'south-africa'])

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df = df.replace({'?':np.NaN, 'y':1, 'n':0,})

dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [4]:
dem.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                         ...                         
democrat                156.0          120.0  ...      160.0         173.0

[1 rows x 16 columns]

In [5]:
rep.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                           ...                         
republican                 31.0           75.0  ...       14.0          96.0

[1 rows x 16 columns]

## What do democrats support more than republicans with p < 0.01

1) Null Hypothesis: 

- No difference in votes on handicapped infants

$H_0$:  $\bar{x}_1 = \bar{x}_2$

2) Alternative Hypothesis

- Democrats voted in support more than republicans

$H_a$: $\bar{x} \neq 0$

3) Confidence Level: 99.9% or 0.999

In [6]:
# t-testing democrats vs republicans based on 'handicapped-infants'
ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

4) t-statistic: 9.205264294809222

5) p-value of 1.613440327937243e-18

--- 

Conclusion: Due to a p-value of 0 I reject the null hypothesis that there was no difference in vote count between Democrats and Republicans.

In [7]:
dem['handicapped-infants'].mean()

0.6046511627906976

In [8]:
rep['handicapped-infants'].mean()

0.18787878787878787

## What do republicans support more than democrats with p < 0.01

1) Null Hypothesis: 

- No difference in votes on Physician fee freeze

$H_0$:  $\bar{x}_1 = \bar{x}_2$

2) Alternative Hypothesis

- Republicans voted in support more than Democrats

$H_a$: $\bar{x} \neq 0$

3) Confidence Level: 99.9% or 0.999

In [9]:
# t-testing republicans vs democrats based on 'physician-fee-freeze', and who had the hight % to vote
ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

4) t-statistic: 49.36708157301406

5) p-value of 1.994262314074344e-177

--- 

Conclusion: Due to a p-value of 0 I reject the null hypothesis that there was no difference in vote count between Democrats and Republicans.

In [12]:
rep['physician-fee-freeze'].mean()

0.9878787878787879

In [13]:
dem['physician-fee-freeze'].mean()

0.05405405405405406

## Where is the difference between republicans and democrats with p > 0.1

1) Null Hypothesis: 

- No difference in votes on water project

$H_0$:  $\bar{x}_1 = \bar{x}_2$

2) Alternative Hypothesis

- Republican and Democrats both voted roughly equally on this topic

$H_a$: $\bar{x} \neq 0$

3) Confidence Level: 99.0% or 0.99

In [14]:
ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

4) t-statistic: 0.08896538137868286

5) p-value of 0.9291556823993485

--- 

Conclusion: Due to a p-value of > 0.1 we failed to reject the null hypothesis.

In [15]:
dem['water-project'].mean()

0.502092050209205

In [16]:
rep['water-project'].mean()

0.5067567567567568

## testing 1-sample t-tests

In [36]:
# Republicans and crime
ttest_1samp(rep['crime'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=91.79687721631203, pvalue=2.672177432250972e-140)

In [41]:
# Democrats and crime
ttest_1samp(dem['crime'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=11.745810821577512, pvalue=9.087409645908879e-26)